Feature/hra

bzhr opened the pull request:
## First version of HRA collector.
- The collector starts with date today if no entries are found in the database and requests data in the amount of days specified in `query_period`. 
- If there's data in the table the collector checks the earliest date requested from the API and starts with one day before that.
- Every `application_id` is queried before insertion and if this `application_id` is already in the table it is not inserted to keep the latest version of that application.

bzhr committed:
Added code for filling warehouse which fails with a weird Postgresql Error

bzhr committed:
Added Record.ensure as False

bzhr committed:
First version of the collector to get historic data from the API

bzhr committed:
Added a check to find if application_ already in database before insertion

pwalsh commented:
two general points:
1. tests?
2. travis is failing

pwalsh commented:
Also @BenGosub you have committed files created by your editor: please remove them from the history. example: `collector.py~`

pwalsh commented on collectors/hra/collector.py:
```diff
@@ -0,0 +1,140 @@
+
+parse_datetime = base.helpers.parse_datetime
+
+url = 'https://stage.harp.org.uk/HARPApiExternal/api/ResearchSummaries'
```
why is this not an env var?

pwalsh commented on collectors/hra/collector.py:
```diff
@@ -0,0 +1,140 @@
+    # 1 - returns all studies for the given period
+    # 2 - returns only studies published in the period
+    # 3 - returns only studies modified in the period
+    result_str = url + '?datePublishedFrom=' + str(from_) + '&datePublishedTo=' + str(to)
```
use python string formatting methods (`''.format()`), not `+`, for string formatting.

pwalsh commented on collectors/hra/collector.py:
```diff
@@ -0,0 +1,140 @@
+        for application in response.json():
+            try:
+                data = {}
+                data['last_updated'] = str(date.today())
```
Use of `str` like this is not py 2 and 3 compatible ( https://github.com/okfn/coding-standards#python-23 ). 

@roll @vitorbaptista how are we dealing with `compat` in this codebase?

roll replied:
https://pythonhosted.org/six/

Here I suppose the best way to use `datetime.date` methods.

bzhr replied:
I've added `date.strftime('%Y/%m/%d')` for date conversions.

For integers `str(int)` is py2/3 compatible right?

vitorbaptista replied:
Do we need to make this code py2/3 compatible? Couldn't we just say we're using Python3.5 and be done with it? This code isn't meant for "public consumption", so we can control where it should run.

roll replied:
We had a discussion about it (I had the same question/proposal) with summary - OKI needs to provide py2/3 support for all open-sourced projects to make it wide deployable. An irony here is anyway we can't do this because `scrapy` and `docker-cloud` are still py2 only.

I have an issue related to it - https://github.com/opentrials/opentrials/issues/110

roll replied:
@BenGosub 
py2/py3 comp - `six.u(str(int))` or `u'%s' % int`

pwalsh replied:
@vitorbaptista py2/3 support is an investment for the future. I'd love to go all Py3 on **new** projects, but not always possible, as @roll points out above.

vitorbaptista replied:
Fair enough, forgot about Scrapy :/

pwalsh commented on collectors/hra/collector.py:
```diff
@@ -0,0 +1,140 @@
+    if (len(response.json()) > 0):
+        while True:
+            to_date = get_to_date()
+            print(to_date)
```
Do we need to use `print`?

bzhr committed:
Fixed screipt for pep8 compliance

pwalsh commented:
@roll i checked some basic stuff, but can you review please.

roll commented:
@pwalsh 
OK, I've assigned me.

roll commented on collectors/hra/collector.py:
```diff
@@ -0,0 +1,137 @@
+                data['research_summary_suitability'] = application['ResearchSummarySuitability']
+
+                # Insert into database only if app_id not inserted already
+                check_id = conn['warehouse'].query("select count(*) from hra where application_id='%s'" % str(application['ApplicationID']))
```
I suppose here could be used `sqlalchemy` parameters.

bzhr replied:
you mean to execute the raw sql statement through sqlalchemy?

roll replied:
I mean something like this instead of string interpolation:

```
conn['warehouse'].query('select count(*) from hra where application_id=?', str(application['ApplicationID']))
```

http://docs.sqlalchemy.org/en/rel_0_9/core/connections.html#sqlalchemy.engine.Connection.execute

bzhr replied:
@roll I run:
`check_id = conn['warehouse'].query('select count(*) from hra where application_id=?', app_id)`
where `app_id` is `unicode 311341`
this fails with:
`TypeError: query() takes exactly 2 arguments (3 given)`

Database query method not the same as SQLALchemy's execute method?

roll replied:
https://dataset.readthedocs.io/en/latest/api.html#dataset.Database.query

Sorry it looks like only keyword params could be used (please check not tested):

```
conn['warehouse'].query('select count(*) from hra where application_id=:app_id', app_id=app_id)
```

bzhr replied:
it works!

roll commented:
@BenGosub 
I'll be able to run this code tomorrow morning to finish a review. For now some things:
- if you use `#Module API` comment on top of file you should put helper function `make_request` to bottom of the file as `_make_request` (with comment `#Internal` or similiar above). Also there are other function which could be written this way instead of inside `collect`.
- function `parse_response` should be moved to `parser.py` file I suppose like the way used in other collectors (with write functionality removed from this function). 
- some log message are missing about collector progress (see other `collectors`)

PS.
Don't forget `make test`.

roll commented on collectors/hra/collector.py:
```diff
@@ -0,0 +1,137 @@
+        while True:
+            to_date = get_to_date()
+            from_date = to_date - timedelta(days=query_period)
+            print('Sleeping for 30sec.')
```
Please use `logger` instead of `print`

bzhr replied:
> function parse_response should be moved to parser.py file I suppose like the way used in other collectors (with write functionality removed from this function).

@roll Where should I put the write functionality?

roll replied:
I mean if you prefer to have separate function for it (we could write this `data['']=` just inside our cycle without function) than is should be something like:

> parser.py

```
def parse_records(response):
   records = []
   # fill records
   return records
```

and then in `collect` write this records. So function with name `parse` will be parsing not parsing and writing.

bzhr committed:
Made correctionsto: remove .py~ files, py2/3 compatible string conversion, proper structure of project, added parsing funct. in parser.py

bzhr committed:
Made correctionsto: remove .py~ files, py2/3 compatible string conversion, proper structure of project, added parsing funct. in parser.py

bzhr committed:
Moved write functionality to collector

bzhr commented:
# Update

@roll 
- Changed the URL with an environmental variable.
- Date to string conversion is through `strftime`; Integer conversion through `six` package.
- Removed .py~ files from commit history.
- Proper string formating.
- Write functionality is in `collector.py`, parsing is in `parser.py`.
- Moved internal functions below to `# Internal` and these are now named with trailing `_`.

bzhr committed:
fixed indentation for code in _get_to_date function

bzhr committed:
fixed string formatting in _make_request function

bzhr committed:
Removed unused variable - parse_datetime

roll commented:
@BenGosub 
About `parse_response` - correct me if I'm wrong but it looks like code just returns on first iteration of `for-in` cycle instead of yielding all records (if response.json() has many applications).

roll commented on collectors/hra/parser.py:
```diff
@@ -0,0 +1,75 @@
+            data = {}
+            data['last_updated'] = date.today().strftime('%Y-%m-%d')
+            data['api_date_from'] = from_date.strftime('%Y-%m-%d')
+            data['api_date_to'] = to_date.strftime('%Y-%m-%d')
```
I'm not fully into HRA API for now but curious why we need to store request attributes as record attributes? If not it also could be removed from function args.

bzhr replied:
> About parse_response - correct me if I'm wrong but it looks like code just returns on first iteration of for-in cycle instead of yielding all records (if response.json() has many applications).

@roll True, I haven't been very careful here.

bzhr replied:
> I'm not fully into HRA API for now but curious why we need to store request attributes as record attributes? If not it also could be removed from function args.

@roll 
The API takes `From` and `To` dates that aren't present in the data. As I understand these dates are used to request data that has been added/updated in the database in that period. I wanted to have something handy for when the collector restarts, can we log `From Date` instead, would this be useful? I will have a look at `Updated Date` and `Publication Date` and see if these are close enough to `From Data` so that when restarting the collector will start according to the `Updated Date` field (with some offset so that it won't miss data) .

BenGosub replied:
# Update

@roll 
- Collector starts at where first records is found (I found this to be Decembar 2008, I checked till 1984 and there are no other records)
- This simplifies the logic significantly, no need to keep parts of the endpoint, less checks, there's only one loop.
- Parse script is only parsing functionality, only response is passed to this function.
- On restart the collector will query the database for the latest `updatedDate` and begin from there.

roll commented on collectors/hra/parser.py:
```diff
@@ -0,0 +1,75 @@
+logger = logging.getLogger(__name__)
+
+
+def parse_response(response, endpoint, from_date, to_date, errors, success, conn):
```
Ideally we don't need args like errors, success, conn here. Also could `endpoint` be extracted from `response`?

bzhr replied:
I am having a look at the moment at how can publication/updated date from the API be used to formulate start date of the endpoint.

bzhr replied:
I suppose you mean I should keep `parser.py` only for parsing functionality, which seems right... It was written like this because I just copied my parsing function that I had written when everything was in `collecot.py`. I will move logging back to `collector.py` and also no need to pass the `conn`.

roll commented:
@BenGosub 
about tests - it's simple to implement a few high-level checks like:

> tests/test_hra.py

``` python
conf = {}
conn = {'warehouse': dataset.connect('sqlite:///:memory:')}
# call collect(conf, conn) - it looks like we need here to introduce some records_limit arg
# check conn['warehouse']['hra'] table data
```

bzhr committed:
New version, collector starts at earlist record and collects until today

bzhr commented:
# Update

@roll 
- Collector starts at where first records is found (I found this to be Decembar 2008, I checked till 1984 and there are no other records)
- This simplifies the logic significantly, no need to keep parts of the endpoint, less checks, there's only one loop.
- Parse script is only parsing functionality, only response is passed to this function.
- On restart the collector will query the database for the latest `updatedDate` and begin from there.

roll commented:
@BenGosub 
Great! I'll take a look soon. Sorry I wasn't able to answer all questions because of business on other parts of the project.

roll commented on collectors/hra/parser.py:
```diff
@@ -0,0 +1,60 @@
+logger = logging.getLogger(__name__)
+
+
+def parse_response(application, response, from_date, to_date):
```
It looks like we don't use `from_date` and `to_date` here? It's great but than this args should be removed.

roll commented on collectors/hra/parser.py:
```diff
@@ -0,0 +1,60 @@
+
+def parse_response(application, response, from_date, to_date):
+        data = {}
+        data['last_updated'] = date.today().strftime('%Y-%m-%d')
```
We could use `meta_updated` builtin field instead of this.

bzhr replied:
@roll how can I access meta_updated field? Cannot find an example in the other collectors...

roll replied:
You could see it in database or `base.Record`. So it will be like `'select max(meta_updated) from hra'`.

bzhr replied:
I see, ok.

roll commented:
Code looks good (just a few comments). Tomorrow morning I will be able to run this against HRA to check business logic.

bzhr committed:
Removed extra arguments from parse function

bzhr committed:
Using meta_updated for last updated field, for finding dafe from changed from max(updated_date) to max(published_date)

roll commented:
@BenGosub 
Checking the business logic. Have you implemented a logic about sleep-exit if we can't query HRA because of schedule? If not we could use for example env variable to do not exit on HRA staging (for dev purposes) and exit on HRA production.

https://github.com/opentrials/opentrials/issues/128

pwalsh commented:
@roll 

What do you think is best? Just exit if collector is running outside of HRA hours?

roll commented:
@pwalsh 
@BenGosub 

``` python
if "can't query" and "hra_production":
    log()
    sleep(1h)
    exit()
```

It will be working now and after `restart: always` will be added to this service. We can't exit immediately because it will be restarted by docker but we can have sleep iteration (sleep doesn't use cpu in python). So every 1h this collector will be checking schedule (collect if could).

Also collector could sleep 1h if there is nothing new to collect but I don't things @BenGosub should focus on this exit-sleep things now because I'll be just implementing it later for all collectors. Much more important to write correct datetime logic here (with correct timezone etc).

bzhr committed:
Added sleep/exit logic

bzhr commented:
@roll @pwalsh 
-  Added `HRA_DEVELOPMENT_ENVIRONMENT` env var that is currently `STAGING`.
- Created a datetime object that is equal to `utc_now`.
- Added the collection records logic into a function because it is called twice: once in staging and once if in development and the time is between 6 and 8.
- If it is in `STAGING` collector goes on without further checks.
- If it is not `STAGING` a further check is done if the day is Monday, Wednesday or Thursday and if the time is between 6 and 8. If true collector runs, if false collector sleeps and exits.

roll commented:
Looks good. I suppose `processor` work could be started. I could wait on merging it because from my experience sometimes work on `processor` lead to `collector` changes e.g. split comma separated field and store it as an array in `warehouse` (cc @pwalsh).

roll commented on collectors/hra/collector.py:
```diff
@@ -0,0 +1,104 @@
+    return from_date
+
+
+def _collect_records(conn, url):
```
this code doesn't really work if relation `hra` doesn't exist

bzhr replied:
@roll you mean a specific part/function or the code in general?

roll replied:
Please check for a situation when `hra` table doesn't exist. 

For example this will fail:

```
last_updated_iter = conn['warehouse'].query('select max(meta_updated) from hra')
```

bzhr replied:
when the table doesn't exist `last_updated` should be `date.today()`?

roll replied:
I've dove into it a little bit more and I don't see a reason to have it because you use it only as

``` python
    data['last_updated'] = last_updated.strftime('%Y-%m-%d')
```

but there is auto `meta_updated` for it.

roll replied:
I thought you need `meta_updated` to use in `_get_from_date`..

bzhr replied:
Changes done, removed pause and removed `last_updated` field.

roll commented on collectors/hra/collector.py:
```diff
@@ -0,0 +1,104 @@
+        from_date = to_date + timedelta(days=1)
+        to_date = to_date + timedelta(days=query_period)
+        logger.info('Sleeping for 30sec.')
+        tm.sleep(30)
```
I thought we need around 1 sec sleep here. But even 1 sec could be omitted because we have pause between API calls on data parsing/writing.

bzhr committed:
Removed 30sec. pause and last_updated field
